### PR TITLE
locale refactoring

### PR DESCRIPTION
**Pull request by supertick**

### CODE REVIEW

File: src/main/java/org/myrobotlab/framework/Service.java [hunk 1]
```diff
@@ -156,6 +157,17 @@
   
   protected String serviceVersion = null;
   
+    /**
+   * map of keys to localizations - 
+   * <pre>
+   *  Match Service with current Locale of the Runtime service
+   *  Match Runtime with current Locale of the Runtime service.
+   *  Match Runtime with Default (english) Locale
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/framework/Service.java:165]: I think there should be a check for "Match Service with Default (english) Locale"  prior to falling back to runtime.

File: src/main/java/org/myrobotlab/service/ProgramAB.java [hunk 2]
```diff
@@ -518,6 +516,34 @@ public void setProcessOOB(boolean processOOB) {
     getSession().processOOB = processOOB;
   }
 
+  /**
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/ProgramAB.java:519]: These following changes seem unrelated to the localization support.

File: src/main/java/org/myrobotlab/service/ProgramAB.java [hunk 3]
```diff
@@ -929,17 +955,17 @@ public SearchPublisher getSearch() {
 
   @Override
   public void setLocale(String code) {
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/ProgramAB.java:957]: Setting locale on the ProgramAB service changes it for the entire platform? this surly can't be correct.

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java [hunk 4]
```diff
@@ -570,18 +567,25 @@ public void unsetWakeWord() {
   }
 
   @Override
-  public void setLocale(String code) {
-    locale = new Locale(code);
+  public String getLanguage() {
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java:570]: The intention of this method is to tell you what language the speech recognizer is detecting.  it's is not the same as Runtime locale.   This seems wrong to me.

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java [hunk 5]
```diff
@@ -570,18 +567,25 @@ public void unsetWakeWord() {
...
-    return locale.getLanguage();
+  public Locale getLocale() {
+    return Runtime.getInstance().getLocale();
   }
 
   @Override
-  public Locale getLocale() {
-    return locale;
+  public Map<String, Locale> getLocales() {
+    return Locale.getAvailableLanguages();
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java:581]: This is wrong, this should be the set of locales that the speech recognizer actually supports.  This should be abstract and force the subclasses to return the list of supported locales.

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java [hunk 6]
```diff
@@ -570,18 +567,25 @@ public void unsetWakeWord() {
...
+  }
+
+  @Override
+  public void setLocale(String code) {
+    // locale = new Locale(code);
+    Runtime.getInstance().setLocale(code);
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java:587]: Setting the locale on this service should not change locale for the platform.

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechSynthesis.java [hunk 7]
```diff
@@ -1068,17 +1065,17 @@ public boolean isMute() {
...
   @Override
   public void setLocale(String code) {
-    locale = new Locale(code);
+    Runtime.getInstance().setLocale(code);
   }
 
   @Override
   public String getLanguage() {
-    return locale.getLanguage();
+    return Runtime.getInstance().getLanguage();
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechSynthesis.java:1073]: This should return the language associated with the currently chosen voice.  not runtime locale

File: src/main/java/org/myrobotlab/service/data/Locale.java [hunk 8]
```diff
@@ -162,10 +166,16 @@ public String getDisplayCountry() {
...
   public String toString() {
     return getTag();
+  } 
+
+  public static Properties loadLocalizations(String fullPath) {
+    Properties props = new Properties();
+    try {
+      props.load(new InputStreamReader(new FileInputStream(fullPath), Charset.forName("UTF-8")));
+    } catch(Exception e) {
+      /* don't care common use case */
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/data/Locale.java:176]: There should at a minimum be a log message here with the exception.

File: src/main/java/org/myrobotlab/service/ProgramAB.java (hunk 3 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/ProgramAB.java:957]: it's preserving the same interface as before, so consider it a helper - or consider the back end has finally been normalized

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java (hunk 4 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java:570]: no .. its original intention was to support the Locale.  And its interface supported changing and getting Locale information including language

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java (hunk 5 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java]: agreed - fixed

File: src/main/java/org/myrobotlab/framework/Service.java (hunk 1 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/framework/Service.java:171]: done ..

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechSynthesis.java (hunk 7 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechSynthesis.java:1073]: it supports the LocaleProvider interface - a setting getting voice operates on different data

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java (hunk 6 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java:582]: I believe it should - the details of hierarchy of locales over the hierarchy of properties for multiple locales - first we need to make it work as best as possible with a single Locale .. mutliple Locales can be addressed in future functionality

File: src/main/java/org/myrobotlab/service/data/Locale.java (hunk 8 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/data/Locale.java:176]: I'll update it to an a debug, but its not an error or even a warning

File: src/main/java/org/myrobotlab/service/ProgramAB.java (hunk 2 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/ProgramAB.java:519]: one very small addition - I don't want to split up ProgramAB into parts for this commit - originally I was working on ProgramAB, before I got squirreled with localize

File: src/main/java/org/myrobotlab/framework/Service.java (hunk 1 above)
Comment (kwatters) [reply src/main/java/org/myrobotlab/framework/Service.java:171]: Javadoc wasn't updated...

File: src/main/java/org/myrobotlab/framework/Service.java [hunk 9]
```diff
@@ -2842,5 +2872,67 @@ public boolean isDev() {
...
+  public String localize(String key, Object ... args) {
+     if (key == null) {
+       log.error("localize(null) not allowed");
+       return null;
+     }     
+     key = key.toUpperCase();     
+     Object prop = localization.get(key);
+     
+     if (prop == null) {
+       prop = defaultLocalization.get(key);
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/framework/Service.java:2903]: Will defaultLocalization ever be null?  I think you need to check that.. no?

File: src/main/java/org/myrobotlab/service/ProgramAB.java [hunk 10]
```diff
@@ -929,17 +955,17 @@ public SearchPublisher getSearch() {
...
   @Override
   public void setLocale(String code) {
-    this.locale = new Locale(code);
+    Runtime.getInstance().setLocale(code);
   }
 
   @Override
   public String getLanguage() {
-    return locale.getLanguage();
+    return Runtime.getInstance().getLanguage();
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/ProgramAB.java:963]: This should be the language from the bot that is currently loaded.  I don't think this is right.

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java (hunk 4 above)
Comment (kwatters) [reply src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java:570]: This should be the language that the speech recognizer is recognizing in.  I don't think this is expected behavior.

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java (hunk 6 above)
Comment (kwatters) [reply src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java:582]: There is no validation that you can set the locale of the speech recognizer that the speech recognizer supports it.  This is very misleading...

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechSynthesis.java (hunk 7 above)
Comment (kwatters) [reply src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechSynthesis.java:1073]: Remove the interface then.  This is wrong.  Ifyou set the langauge of speech synthesis, it should choose a voice that supports the desired langauge.. or thow and exception if the speech synthesis doesn't support that language.  This is very confusing now.

File: src/main/java/org/myrobotlab/service/data/Locale.java (hunk 8 above)
Comment (kwatters) [reply src/main/java/org/myrobotlab/service/data/Locale.java:176]: Why woulnd't it be an error message?  if the file system croaks on you.. this would throw an IO exception.A log message should be here.  Also.. what is closing the Input stream readers, aren't you leaking file handles here?

File: src/main/java/org/myrobotlab/framework/interfaces/ServiceInterface.java [hunk 11]
```diff
@@ -134,4 +134,7 @@
 
   public String getFullName();
 
+  public void loadLocalizations();
+  public void loadLocalizations(String tag);
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/framework/interfaces/ServiceInterface.java:138]: "tag" should be called "language" or "locale"  it doesn't make sense to call it "tag"

File: src/main/java/org/myrobotlab/service/GoogleSearch.java [hunk 12]
```diff
@@ -94,7 +92,8 @@ public void clearFilters() {
 
   @Override
   public void setLocale(String code) {
-    locale = new Locale(code);
+    // locale = new Locale(code);
+    Runtime.getInstance().setLocale(code);
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/GoogleSearch.java:96]: You should be able to set the language of your search.  This should not be setting the locale of runtime.

File: src/main/java/org/myrobotlab/service/GoogleSearch.java [hunk 13]
```diff
@@ -126,7 +125,7 @@ public SearchResults search(String searchText) throws IOException {
       String encodedSearch = URLEncoder.encode(searchText, "UTF-8");
 
       // not sure if locale is supported tag probably is ....
-      String request = "https://google.com/search?lr=lang_" + locale.getTag() + "&q=" + encodedSearch + "&aqs=chrome..69i57.5547j0j7&sourceid=chrome&ie=UTF-8";
+      String request = "https://google.com/search?lr=lang_" + Runtime.getInstance().getLocaleTag() + "&q=" + encodedSearch + "&aqs=chrome..69i57.5547j0j7&sourceid=chrome&ie=UTF-8";
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/GoogleSearch.java:128]: This should be using the locale of the google search service.. not runtime.

File: src/main/java/org/myrobotlab/service/GoogleSearch.java [hunk 14]
```diff
@@ -229,7 +228,7 @@ public SearchResults search(String searchText) throws IOException {
     try {
       // can only grab first 100 results
 
-      String url = "https://www.google.com/search?lr=lang_" + locale.getTag() + "&site=imghp&tbm=isch&source=hp&q=" + searchText + "&gws_rd=cr";
+      String url = "https://www.google.com/search?lr=lang_" + Runtime.getInstance().getLocaleTag() + "&site=imghp&tbm=isch&source=hp&q=" + searchText + "&gws_rd=cr";
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/GoogleSearch.java:231]: this should not be from runtime.. this should be the locale of the google search service.   you should be able to search in french even if your current runtime locale is english.

File: src/main/java/org/myrobotlab/service/GoogleSearch.java [hunk 15]
```diff
@@ -385,12 +384,12 @@ public static void main(String[] args) {
 
   @Override
   public String getLanguage() {
-    return locale.getLanguage();
+    return Runtime.getInstance().getLanguage();
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/GoogleSearch.java:387]: This means you can't search in any language other than the current runtime language.

File: src/main/java/org/myrobotlab/service/GoogleSearch.java [hunk 16]
```diff
@@ -385,12 +384,12 @@ public static void main(String[] args) {
...
   @Override
   public String getLanguage() {
-    return locale.getLanguage();
+    return Runtime.getInstance().getLanguage();
   }
 
   @Override
   public Locale getLocale() {
-    return locale;
+    return Runtime.getInstance().getLocale();
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/GoogleSearch.java:392]: Should be locale of the google search service..  if that's null, then fallback to runtime locale.

File: src/main/java/org/myrobotlab/service/ProgramAB.java (hunk 10 above)
Comment (kwatters) [reply src/main/java/org/myrobotlab/service/ProgramAB.java:963]: This language should be derived from the locale property setting for the current bot.   not from runtime.

File: src/main/java/org/myrobotlab/service/ProgramAB.java [hunk 17]
```diff
@@ -929,17 +955,17 @@ public SearchPublisher getSearch() {
...
   }
 
   @Override
   public Locale getLocale() {
-    return locale;
+    return Runtime.getInstance().getLocale();
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/ProgramAB.java:968]: should be locale of the currently loaded bot.

File: src/main/java/org/myrobotlab/service/Runtime.java [hunk 18]
```diff
@@ -2697,6 +2698,12 @@ public static Double getBatteryLevel() {
   @Override
   public void setLocale(String code) {
     locale = new Locale(code);
+    // attempt to reload all services localizations
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/Runtime.java:2701]: This will override the localization for every service..  if you call setLocale at the runtime level,  then you'll have to go back through and reset the locale on speech recognition and on speech synthesis to ensure that those services are still operating in the proper language.

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java [hunk 19]
```diff
@@ -5,6 +5,7 @@
 import java.util.Set;
 import java.util.TreeSet;
 
+import org.apache.tools.ant.taskdefs.LoadProperties;
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java:8]: Seems like a strange library to be importing,, is there a better import that is more generic?

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechSynthesis.java (hunk 7 above)
Comment (kwatters) [reply src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechSynthesis.java:1059]: Get language should return the current language that the current voice is for.  This is returning the runtime langauge which is incorrect.

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechSynthesis.java [hunk 20]
```diff
@@ -1069,16 +1050,18 @@ public boolean isMute() {
...
   }
 
   @Override
   public Locale getLocale() {
-    return locale;
+    return Runtime.getInstance().getLocale();
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechSynthesis.java:1064]: This should be the langauge of the current voice, not runtime.

File: src/main/java/org/myrobotlab/service/InMoov2.java [hunk 21]
```diff
@@ -640,11 +629,16 @@ public InMoov2Arm getLeftArm() {
...
+    return Runtime.getInstance().getLanguage();
+  }
 
-	@Override
-	public Locale getLocale() {
-		return locale;
-	}
+  @Override
+  public Locale getLocale() {
+    return Runtime.getInstance().getLocale();
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/InMoov2.java:640]: This should be the locale of the InMoov2 service.. not Runtime.

File: src/main/java/org/myrobotlab/programab/BotInfo.java [hunk 22]
```diff
@@ -28,32 +29,41 @@
...
     if (bot == null) {
       // lazy loading of bot - created on the first use
-      
       if (properties.containsKey("locale")) {
         bot = new Bot(name, path.getAbsolutePath(), java.util.Locale.forLanguageTag(properties.get("locale")));
       } else {
         bot = new Bot(name, path.getAbsolutePath());
       }
 
+      // merge properties - potentially there are 2 sets
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/programab/BotInfo.java:57]: Where do these other properties come from and how are they managed?  This seems unrelated to localization.

File: src/main/java/org/myrobotlab/service/InMoov2.java [hunk 23]
```diff
@@ -640,11 +629,16 @@ public InMoov2Arm getLeftArm() {
 	public InMoov2Hand getLeftHand() {
 		return leftHand;
 	}
+	
+  @Override
+  public String getLanguage() {
+    return Runtime.getInstance().getLanguage();
```
Comment (kwatters) [inline src/main/java/org/myrobotlab/service/InMoov2.java:635]: This should be the language that the InMoov2 is set to.. not Runtime's language.

File: src/main/java/org/myrobotlab/framework/Service.java (hunk 1 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/framework/Service.java]: updated

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java (hunk 4 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java]: updated - pushed to Service - each service has its own Locale

File: src/main/java/org/myrobotlab/framework/Service.java (hunk 9 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/framework/Service.java:2910]: no .. it will not. The Service constructor loads localization and default if null will always be initialized to a map - even if en.properties is not found

File: src/main/java/org/myrobotlab/service/ProgramAB.java (hunk 10 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/ProgramAB.java]: this has been pushed up to Service .. and all services use their local property - if not set then its set from runtime

File: src/main/java/org/myrobotlab/framework/interfaces/ServiceInterface.java (hunk 11 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/framework/interfaces/ServiceInterface.java]: its been removed

File: src/main/java/org/myrobotlab/service/GoogleSearch.java (hunk 12 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/GoogleSearch.java]: been completely removed and pushed up into Service

File: src/main/java/org/myrobotlab/service/GoogleSearch.java (hunk 13 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/GoogleSearch.java]: nice catch ! .. thanks ..
fixed

File: src/main/java/org/myrobotlab/service/GoogleSearch.java (hunk 15 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/GoogleSearch.java]: good catch .. these needed to be removed

File: src/main/java/org/myrobotlab/service/GoogleSearch.java (hunk 16 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/GoogleSearch.java]: fixed - taken care of by parent Service

File: src/main/java/org/myrobotlab/service/ProgramAB.java (hunk 17 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/ProgramAB.java]: removed .. promoted .. normalized - no longer an issue

File: src/main/java/org/myrobotlab/service/Runtime.java (hunk 18 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/Runtime.java]: uses Service.setLocale .. no longer an issue

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java (hunk 19 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechRecognizer.java]: dunno how that got pulled in .. but all this code is "Outdated" - no longer an import

File: src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechSynthesis.java (hunk 20 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/abstracts/AbstractSpeechSynthesis.java]: promoted normalized updated - no longer an issue

File: src/main/java/org/myrobotlab/service/InMoov2.java (hunk 21 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/InMoov2.java]: wow .. i can't believe I left this in .. thanks - again .. nice catch.. removed

File: src/main/java/org/myrobotlab/programab/BotInfo.java (hunk 22 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/programab/BotInfo.java:57]: these are not the properties your looking for ... a programab pr will be coming up .. let it go...

File: src/main/java/org/myrobotlab/service/InMoov2.java (hunk 23 above)
Comment (supertick) [reply src/main/java/org/myrobotlab/service/InMoov2.java]: it is .. should be outdated .. check now
